### PR TITLE
Show pipeline details embedded in runs

### PR DESCRIPTION
**Pull request by yebrahim**
Runs created in notebooks (or created the same way using API/CLI) do not contain a pipeline ID reference. Instead, they contain the entire pipeline spec embedded inside the run object. This PR handles extracting the pipeline details from there and showing them in the `PipelineDetails` view.

Some other small fixes:
- Moved `QUERY_PARAMS` to Router component to make all routing definitions in one place.
- Changed pipeline id to be an optional parameter for the `PipelineDetails` page.
- Changed the custom renderer methods on `RunList` to be public. I think this makes more sense, since they really are part of the public API, and can be called as such.

Part of #75.

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/pipelines/447)

### CODE REVIEW

File: frontend/src/components/Router.tsx [hunk 1]
```diff
@@ -44,6 +44,17 @@ const css = stylesheet({
   },
 });
 
+export enum QUERY_PARAMS {
```
Comment (rileyjbauer) [inline frontend/src/components/Router.tsx]: :+1:

File: frontend/src/components/Router.tsx [hunk 2]
```diff
@@ -58,7 +69,7 @@ export const RoutePage = {
   NEW_EXPERIMENT: '/experiments/new',
   NEW_RUN: '/runs/new',
   PIPELINES: '/pipelines',
-  PIPELINE_DETAILS: `/pipelines/details/:${RouteParams.pipelineId}`,
+  PIPELINE_DETAILS: `/pipelines/details/:${RouteParams.pipelineId}?`, // pipelineId is optional
```
Comment (rileyjbauer) [inline frontend/src/components/Router.tsx]: This `?` is correct? shouldn't it just be added later if there are query params?

File: frontend/src/lib/RunUtils.ts [hunk 3]
```diff
@@ -29,6 +29,10 @@ function getPipelineId(run?: ApiRun | ApiJob): string | null {
   return run && run.pipeline_spec && run.pipeline_spec.pipeline_id || null;
 }
 
+function getPipelineSpec(run?: ApiRun | ApiJob): string | null {
+  return run && run.pipeline_spec && run.pipeline_spec.workflow_manifest || null;
```
Comment (rileyjbauer) [inline frontend/src/lib/RunUtils.ts]: nit: add parentheses?
 return (run && run.pipeline_spec && run.pipeline_spec.workflow_manifest) || null;

File: frontend/src/pages/PipelineDetails.test.tsx [hunk 4]
```diff
@@ -89,7 +93,12 @@ describe('PipelineDetails', () => {
     deletePipelineSpy.mockReset();
     getPipelineSpy.mockImplementation(() => Promise.resolve(testPipeline));
     getPipelineSpy.mockClear();
-    getTemplateSpy.mockImplementation(() => Promise.resolve('{}'));
+    getRunSpy.mockImplementation(() => Promise.resolve(testRun));
```
Comment (rileyjbauer) [inline frontend/src/pages/PipelineDetails.test.tsx]: Are you just using `Promise.resolve` to make it closer to the actual implementation? (i.e. returning a promise?)

File: frontend/src/pages/PipelineDetails.tsx [hunk 5]
```diff
@@ -111,8 +113,9 @@ class PipelineDetails extends Page<{}, PipelineDetailsState> {
   }
 
   public getInitialToolbarState(): ToolbarProps {
+    const fromRunId = new URLParser(this.props).get(QUERY_PARAMS.fromRunId);
```
Comment (rileyjbauer) [inline frontend/src/pages/PipelineDetails.tsx]: why not just do an `if`, and then have two `return`s ? Seems like it'd be more straightforward

File: frontend/src/pages/PipelineDetails.tsx [hunk 6]
```diff
@@ -236,48 +242,91 @@ class PipelineDetails extends Page<{}, PipelineDetailsState> {
...
+        } else {
+          breadcrumbs.push(
+            { displayName: 'All runs', href: RoutePage.RUNS }
+          );
+        }
+        breadcrumbs.push({
+          displayName: runDetails.run!.name!,
+          href: RoutePage.RUN_DETAILS.replace(':' + RouteParams.runId, fromRunId),
+        });
+        pageTitle = 'Pipeline details';
```
Comment (rileyjbauer) [inline frontend/src/pages/PipelineDetails.tsx]: nit: maybe `Embedded Pipeline details`
or
`Embedded Pipeline` ?

The word doesn't have to be "embedded" either

File: frontend/src/pages/PipelineDetails.tsx [hunk 7]
```diff
@@ -151,77 +157,77 @@ class PipelineDetails extends Page<{}, PipelineDetailsState> {
...
+                    </div>}
+                  </div>
+                </SidePanel>
+                <div className={css.footer}>
+                  {!summaryShown && (
+                    <Button onClick={() => this.setStateSafe({ summaryShown: !summaryShown })} color='secondary'>
+                      Show summary
+                      </Button>
+                  )}
+                  <div className={classes(commonCss.flex, summaryShown && css.footerInfoOffset)}>
```
Comment (rileyjbauer) [inline frontend/src/pages/PipelineDetails.tsx]: `summaryShown && css.footerInfoOffset` should either have an additional boolean for when there is no summary (for embedded pipelines), or `summaryShown` should just be set to false in that case. Maybe in `load()` ?

File: frontend/src/pages/RunList.tsx [hunk 8]
```diff
@@ -176,7 +179,101 @@ class RunList extends React.Component<RunListProps, RunListState> {
...
+      return <div>-</div>;
+    }
+    const search = new URLParser(this.props).build({ [QUERY_PARAMS.fromRunId]: id });
+    const url = pipelineInfo.showLink ?
+      RoutePage.PIPELINE_DETAILS.replace(':' + RouteParams.pipelineId + '?', '') + search :
+      RoutePage.PIPELINE_DETAILS.replace(':' + RouteParams.pipelineId, pipelineInfo.id || '');
+    return (
+      <Link className={commonCss.link} onClick={(e) => e.stopPropagation()}
+        to={url}>
+        {pipelineInfo.showLink ? 'View pipeline' : pipelineInfo.displayName}
```
Comment (rileyjbauer) [inline frontend/src/pages/RunList.tsx]: nit: I think maybe `[View pipeline]` would be better? something to help distinguish it visually from a normal pipeline name?

Or maybe we don't actually care about that

File: frontend/src/components/Router.tsx (hunk 2 above)
Comment (yebrahim) [reply frontend/src/components/Router.tsx]: This is the route path, the current location page is matched against all of these paths to see which one matches, so it needs to include the `?`.

File: frontend/src/pages/PipelineDetails.test.tsx (hunk 4 above)
Comment (yebrahim) [reply frontend/src/pages/PipelineDetails.test.tsx]: There's no real need for it here. The only difference I can think of is that this catches the case where the calling code mistakenly doesn't use `await` when calling this function. I realize this isn't the case in all tests though, but I think this is more correct. wdyt?

File: frontend/src/pages/PipelineDetails.tsx (hunk 5 above)
Comment (yebrahim) [reply frontend/src/pages/PipelineDetails.tsx]: Done.

File: frontend/src/pages/PipelineDetails.tsx (hunk 6 above)
Comment (yebrahim) [reply frontend/src/pages/PipelineDetails.tsx]: I don't know if that's easily understood by the user. We can change this later if we have better ideas, but for now I think [the spec](https://github.com/kubeflow/pipelines/issues/75#issuecomment-440441381) is fine.

File: frontend/src/pages/RunList.tsx (hunk 8 above)
Comment (yebrahim) [reply frontend/src/pages/RunList.tsx]: I think we should distinguish them too. I'll add @ajayalfred here for follow up, but for now I think we can merge this whenever it's ready until we hear back from him.

File: frontend/src/pages/PipelineDetails.tsx (hunk 7 above)
Comment (yebrahim) [reply frontend/src/pages/PipelineDetails.tsx]: Yup, added a `!!pipeline` condition in there, that was the missing piece to shift it back.

File: frontend/src/pages/PipelineDetails.test.tsx (hunk 4 above)
Comment (rileyjbauer) [reply frontend/src/pages/PipelineDetails.test.tsx]: Is await actually needed for the spies though? If it is, then I think your solution is an adequate one

Comment (yebrahim) [reply frontend/src/pages/PipelineDetails.test.tsx]: The component's code should be calling `await` on the API. If it's not, it's a bug, and this catches that.

File: frontend/src/pages/RunList.tsx (hunk 8 above)
Comment (ajayalfred) [reply frontend/src/pages/RunList.tsx]: I'll check with Jordan and comment back here.

**Pull request merged**